### ISSUE TITLE
Better error message for first-time-users: No 'Access-Control-Allow-Origin' header is present 

### ISSUE THREAD

**Issue by guettli** [hoppscotch/hoppscotch#1376]
**Is your feature request related to a problem? Please describe.**

I try to execute my first GET request and get this result:

![image](https://user-images.githubusercontent.com/414336/101740755-89393f80-3ac9-11eb-896b-eca0a7a85013.png)

I think the first-time-user experience could get improved here.

The hint to F12 vanishes soon, and I think accessing the error message via F12 is not a nice solution.

It would be nice, if the software could tell me directly what went wrong. Then maybe some explanation, and
a way how to work around this.

**Comment by liyasthomas** [comment]
Makes sense. Toast is too small to be noticed. I appreciate if you can suggest a better way to enhance error state ux?

**Comment by AndrewBastin** [comment]
Umm, we could improve the error UX. But the problem we are facing is that browsers don't let the full details of the network error to be given to the scripts. As far as the script is concerned, we are only given a "Network Error" message and it is pretty generic and is thrown for every network related issue, not only the one you mentioned as an example.

Since we can't exactly know what went wrong, we really can't suggest anything on how to fix it.

**Comment by NBTX** [comment]
This isn't a general solution for any CORS policy - but it would allow us to figure out perhaps if we need to show a full-page dialog prompting them to install the browser extension or enable the proxy.

We could perhaps set up an API route that works in every way except by having an Access-Control-Allow-Origin policy to a random domain. That way we could test by first requesting a 204 to some domain (to establish that the device is connected to the internet) and then make a request to our CORS route - if it succeeds, proxy or extension is enabled, if it fails, the error must be due to CORS hence we can show the full-page dialog.

**Comment by guettli** [comment]
@AndrewBastin this is new to me. The script just gets "NetworkError" without a message? Hmm let's focus on this. How could this be changed? Of course this is not an issue of hoppscotch any more, but on the other hand we could make life easier for thousand developers. Who could help here?

**Comment by AndrewBastin** [comment]
I don't think we have any other ways to get it from a browser request stand point.

The solution provided by @NBTX sounds great. But I don't really encourage it because of security concerns and the user not knowing that request is being ran on the background through a proxy.

**Comment by satyam-52** [comment]
This error can be solved altogether by using a CORS proxy which you can create by yourself. It's quite easy actually. You can use the code <a href="https://github.com/Rob--W/cors-anywhere/">here</a>. Simply use these commands:
```
git clone https://github.com/Rob--W/cors-anywhere.git
cd cors-anywhere/
npm install
heroku create
git push heroku master
```

After you've done this just append the URL you want to fetch from to your proxy URL. For example:

```
https://whispering-brushlands-56177.herokuapp.com/https://evilinsult.com/generate_insult.php?lang=en&type=json
```

I hope this helps.

**Comment by AndrewBastin** [comment]
@satyam-52 I don't know how this fits into the conversation here exactly.

Hoppscotch already has a built in proxy that circumvents this issue. You can enable it in the Settings page.

**Comment by satyam-52** [comment]
@AndrewBastin Just re-read the problem, I just misinterpreted. Will delete my comment :)

**Comment by KartikSoneji** [comment]
@AndrewBastin 
> because of security concerns and the user not knowing that request is being ran on the background through a proxy.

Hmm? How is this any different from running similar network requests (google analytics, feature tracking)?
If the request is being run through a proxy, it is because the user (or someone on their behalf) explicitly enabled it.

I think adding a route like `/connectivity-check/?cors=true` to proxyscotch should allow checking both for network connectivity and if CORS is blocked.
As for mixed content, checking `requestUrl.protocol == window.location.protocol` should cover most cases.

Maybe related: #2248

**Comment by imskr** [comment]
Can I work on this?

**Comment by liyasthomas** [comment]
Hi @imskr, feel free to pick up this feature. Do let us know if you need any help.

**Comment by varun-s22** [comment]
Hi @liyasthomas, is this still open?.. can i take it?
also how are we planning to show the message on the client side?
do we show some popup notifications (like Toast, which i read above)?

**Comment by Amit-Morade** [comment]
can I work on this?

**Comment by VineeTagarwaL-code** [comment]
Assign me please , i want to work on this

**Comment by saifkp517** [comment]
has there been any updates? if not could i get assigned to this?

**Comment by guettli** [comment]
@saifkp517 I just created the issue. I won't work on it, since I work in a different environment now. I think: Just do it!

**Comment by JunaidMancheri** [comment]
@liyasthomas Is this task still pending? If yes I would love to work on it. Please let me know, thank you!

**Comment by farhanmir** [comment]
I can work on this if this is still pending.

**Comment by A-yushupadhyay** [comment]
Hi, I’m Ayush — I’d like to work on this issue. Can I take this one?

**Comment by ansonsaju** [comment]
Hi, I’m Anson 👋  
I’m interested in working on this issue as part of my open-source learning.  
May I take this up?

**Comment by saksham242007** [comment]
Hi! I'd like to work on improving this error message and add a clearer explanation in the UI. Could you please assign this to me?

**Comment by VishwaVikas2006** [comment]
Hi, I’d like to work on this issue. Could you please assign it to me?

**Comment by codewithaslam** [comment]
Hi! I would like to work on this issue as part of the FOSS United Hackathon.
Could this issue be assigned to me?

**Comment by nv-2177** [comment]
Hi, I would like to work on this issue.
Since browsers only expose a generic "Network Error", I plan to improve the UX by showing a more helpful message suggesting possible causes like CORS restrictions and recommending enabling the proxy in settings.

Please let me know if this approach works.

**Comment by farheenayy33** [comment]
Hi, I’m a frontend developer and would like to work on improving the error UX for CORS/network errors. Can you please confirm if this is still needed and give any guidelines/preferences for the design?

**Comment by hajraahmad21** [comment]
Hi @guettli I wuld like to work on it

**Comment by Ganesh-Code-Lives** [comment]
Hi, I'd like to work on this issue. Is it still open?

**Comment by puneetdixit200** [comment]
Hi team, I’d like to work on this issue. I can propose and implement a clearer first-time CORS error message flow, then open a PR with tests/docs updates. Please assign this to me if available.

**Comment by kilan080** [comment]
Hi, I'd like to work on this. I'll look into where the CORS error is currently displayed and propose a better message

**Comment by ananikets18** [comment]
Hey there! I'd love to take this on. 

I'm thinking of updating the generic network error text to explicitly mention CORS, and suggesting they switch Interceptors to bypass it. It's a simple change but should really help first-time users figure out what went wrong without cluttering the UI.

Happy to open a PR if this sounds good to you!
